File: packages/java-sdk/src/test/java/com/turbodocx/TurboQuoteScheduleTest.java
package com.turbodocx;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.turbodocx.models.SignatureSchedule;
import com.turbodocx.models.quote.*;
import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;
import okhttp3.mockwebserver.RecordedRequest;
import org.junit.jupiter.api.*;

import java.io.IOException;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

/**
 * TurboQuote reminder + expiration schedule serialization tests.
 *
 * <p>Mirrors the js-sdk, py-sdk, go-sdk and ruby-sdk turboquote-schedule suites, per the
 * cross-SDK test-parity rule.
 *
 * <p>The quote send endpoints are JSON (unlike the multipart signature send), so the eight
 * schedule fields ride FLAT at the top level of the request body — NOT nested under a "schedule"
 * key — and durations serialize as plain {@code {value, unit}} OBJECTS, not JSON-encoded strings.
 * The boxed schedule fields keep a deliberate {@code false} / {@code 0} distinct from "unset": an
 * omitted field stays off the wire and inherits the org default. Request-body keys are camelCase.
 */
class TurboQuoteScheduleTest {

    private MockWebServer server;
    private TurboQuoteClient client;
    private final Gson gson = new Gson();

    @BeforeEach
    void setUp() throws IOException {
        server = new MockWebServer();
        server.start();
        client = new TurboQuoteClient.Builder()
                .apiKey("test-api-key")
                .orgId("test-org-id")
                .baseUrl(server.url("/").toString())
                .build();
    }

    @AfterEach
    void tearDown() throws IOException {
        server.shutdown();
    }

    private String wrapInData(Object body) {
        JsonObject wrapper = new JsonObject();
        wrapper.add("data", gson.toJsonTree(body));
        return gson.toJson(wrapper);
    }

    private void enqueueSendResponse() {
        Map<String, Object> result = new HashMap<>();
        result.put("id", "q-1");
        result.put("status", "sent");
        Map<String, Object> response = new HashMap<>();
        response.put("result", result);
        response.put("message", "Quote sent");
        response.put("documentId", "doc-2");
        server.enqueue(new MockResponse().setBody(wrapInData(response)));
    }

    private void assertDuration(JsonObject body, String key, int wantValue, String wantUnit) {
        assertTrue(body.has(key), key + " was not sent");
        assertTrue(body.get(key).isJsonObject(), key + " should be a {value, unit} object, not a string");
        JsonObject obj = body.getAsJsonObject(key);
        assertEquals(wantValue, obj.get("value").getAsInt(), key + ".value");
        assertEquals(wantUnit, obj.get("unit").getAsString(), key + ".unit");
    }

    @Test
    @DisplayName("sendQuote sends every schedule field flat with object durations")
    void sendQuoteSendsEveryFieldFlat() throws Exception {
        enqueueSendResponse();

        SignatureSchedule schedule = SignatureSchedule.builder()
                .remindersEnabled(true)
                .reminderDelay(new SignatureSchedule.Duration(3, "days"))
                .reminderInterval(new SignatureSchedule.Duration(12, "hours"))
                .maxReminders(5)
                .expirationEnabled(true)
                .expireAfter(new SignatureSchedule.Duration(30, "days"))
                .expirationWarning(new SignatureSchedule.Duration(3, "days"))
                .expirationWarningInterval(new SignatureSchedule.Duration(1, "days"))
                .build();

        SendQuoteRequest request = new SendQuoteRequest();
        request.setCcEmails(Collections.singletonList("admin@example.com"));
        request.setSchedule(schedule);

        client.turboQuote().sendQuote("q-1", request);

        RecordedRequest recorded = server.takeRequest();
        assertTrue(recorded.getPath().endsWith("/v1/quotes/q-1/send"));
        JsonObject body = gson.fromJson(recorded.getBody().readUtf8(), JsonObject.class);

        // Flat at the top level, never nested under "schedule".
        assertFalse(body.has("schedule"), "schedule fields must be FLAT, not nested under \"schedule\"");

        // Native boolean / number — not stringified.
        assertTrue(body.get("remindersEnabled").getAsBoolean());
        assertTrue(body.get("expirationEnabled").getAsBoolean());
        assertEquals(5, body.get("maxReminders").getAsInt());
        assertTrue(body.get("remindersEnabled").getAsJsonPrimitive().isBoolean());
        assertTrue(body.get("maxReminders").getAsJsonPrimitive().isNumber());

        // Durations are objects, not JSON strings.
        assertDuration(body, "reminderDelay", 3, "days");
        assertDuration(body, "reminderInterval", 12, "hours");
        assertDuration(body, "expireAfter", 30, "days");
        assertDuration(body, "expirationWarning", 3, "days");
        assertDuration(body, "expirationWarningInterval", 1, "days");

        assertEquals("admin@example.com", body.getAsJsonArray("ccEmails").get(0).getAsString());
    }

    @Test
    @DisplayName("sendQuote omits every schedule key when unset")
    void sendQuoteOmitsUnsetFields() throws Exception {
        enqueueSendResponse();

        SendQuoteRequest request = new SendQuoteRequest();
        request.setCcEmails(Collections.singletonList("admin@example.com"));
        client.turboQuote().sendQuote("q-1", request);

        RecordedRequest recorded = server.takeRequest();
        JsonObject body = gson.fromJson(recorded.getBody().readUtf8(), JsonObject.class);

        for (String key : new String[]{
                "remindersEnabled", "reminderDelay", "reminderInterval", "maxReminders",
                "expirationEnabled", "expireAfter", "expirationWarning", "expirationWarningInterval"}) {
            assertFalse(body.has(key), key + " should be omitted so the org default applies");
        }
        assertFalse(body.has("schedule"));
    }

    // false and 0 are meaningful, not "unset" — the boxed fields keep them off the org-default path.
    @Test
    @DisplayName("sendQuote preserves the meaningful zeros maxReminders:0 and expirationEnabled:false")
    void sendQuotePreservesMeaningfulZeros() throws Exception {
        enqueueSendResponse();

        SignatureSchedule schedule = SignatureSchedule.builder()
                .maxReminders(0)
                .expirationEnabled(false)
                .build();
        SendQuoteRequest request = new SendQuoteRequest();
        request.setSchedule(schedule);

        client.turboQuote().sendQuote("q-1", request);

        RecordedRequest recorded = server.takeRequest();
        JsonObject body = gson.fromJson(recorded.getBody().readUtf8(), JsonObject.class);

        assertEquals(0, body.get("maxReminders").getAsInt());
        assertFalse(body.get("expirationEnabled").getAsBoolean());
    }

    @Test
    @DisplayName("sendQuoteWithDeliverable carries the schedule flat alongside the deliverable fields")
    void sendQuoteWithDeliverableCarriesScheduleFlat() throws Exception {
        enqueueSendResponse();

        SignatureSchedule schedule = SignatureSchedule.builder()
                .remindersEnabled(true)
                .reminderDelay(new SignatureSchedule.Duration(2, "days"))
                .expirationEnabled(false)
                .build();
        SendQuoteWithDeliverableRequest request = new SendQuoteWithDeliverableRequest();
        request.setDeliverableId("del-1");
        request.setMergePosition("end");
        request.setSchedule(schedule);

        client.turboQuote().sendQuoteWithDeliverable("q-1", request);

        RecordedRequest recorded = server.takeRequest();
        assertTrue(recorded.getPath().endsWith("/v1/quotes/q-1/send-with-deliverable"));
        JsonObject body = gson.fromJson(recorded.getBody().readUtf8(), JsonObject.class);

        assertFalse(body.has("schedule"));
        assertEquals("del-1", body.get("deliverableId").getAsString());
        assertTrue(body.get("remindersEnabled").getAsBoolean());
        assertDuration(body, "reminderDelay", 2, "days");
        assertFalse(body.get("expirationEnabled").getAsBoolean());
    }

    @Test
    @DisplayName("createAndSend emits the flat schedule on the send step")
    void createAndSendEmitsFlatScheduleOnSendStep() throws Exception {
        // Step 1: create quote
        Map<String, Object> quote = new HashMap<>();
        quote.put("id", "q-1");
        quote.put("status", "draft");
        Map<String, Object> createResponse = new HashMap<>();
        createResponse.put("result", quote);
        createResponse.put("message", "Quote created successfully");
        server.enqueue(new MockResponse().setBody(wrapInData(createResponse)));
        // Step 2: send quote
        enqueueSendResponse();

        SignatureSchedule schedule = SignatureSchedule.builder()
                .remindersEnabled(true)
                .maxReminders(0)
                .reminderDelay(new SignatureSchedule.Duration(1, "days"))
                .build();
        SendQuoteRequest send = new SendQuoteRequest();
        send.setSchedule(schedule);

        CreateAndSendRequest request = new CreateAndSendRequest();
        request.setName("Enterprise License");
        request.setCompanyId("c-1");
        request.setContactId("ct-1");
        request.setSend(send);

        client.turboQuote().createAndSend(request);

        RecordedRequest create = server.takeRequest();
        assertTrue(create.getPath().endsWith("/v1/quotes"));
        RecordedRequest sendReq = server.takeRequest();
        assertTrue(sendReq.getPath().endsWith("/v1/quotes/q-1/send"));

        JsonObject body = gson.fromJson(sendReq.getBody().readUtf8(), JsonObject.class);
        assertFalse(body.has("schedule"));
        assertTrue(body.get("remindersEnabled").getAsBoolean());
        assertEquals(0, body.get("maxReminders").getAsInt());
        assertDuration(body, "reminderDelay", 1, "days");
    }
}

File: packages/js-sdk/tests/turboquote-schedule.test.ts
/**
 * TurboQuote reminder + expiration schedule serialization tests.
 *
 * The quote send endpoints are JSON (unlike the signature send, which is multipart), so the eight
 * schedule fields ride FLAT at the top level of the request body — NOT nested under a "schedule"
 * key — and durations serialize as plain `{ value, unit }` OBJECTS, not JSON-encoded strings.
 * Presence is null-checked, so a deliberate `false` / `0` survives while an unset field is omitted
 * and inherits the org default. Request-body keys stay camelCase.
 *
 * This mirrors how the signature-schedule suite inspects the mocked HTTP request body — here the
 * body is the second argument the SDK hands to `HttpClient.post`.
 */

import { TurboQuote } from "../src/modules/quote";
import { HttpClient } from "../src/http";

jest.mock("../src/http", () => {
  const actual = jest.requireActual("../src/http");
  return {
    ...actual,
    HttpClient: jest.fn(),
  };
});

const MockedHttpClient = HttpClient as jest.MockedClass<typeof HttpClient>;

const SCHEDULE_KEYS = [
  "remindersEnabled",
  "reminderDelay",
  "reminderInterval",
  "maxReminders",
  "expirationEnabled",
  "expireAfter",
  "expirationWarning",
  "expirationWarningInterval",
] as const;

describe("TurboQuote schedule serialization", () => {
  let mockClient: { post: jest.Mock };

  beforeEach(() => {
    jest.clearAllMocks();
    (TurboQuote as any).client = undefined;
    mockClient = { post: jest.fn() };
    MockedHttpClient.mockImplementation(() => mockClient as any);
    TurboQuote.configure({ apiKey: "test-key", orgId: "org-1" });
  });

  describe("sendQuote — JSON send path", () => {
    it("sends every schedule field FLAT at the top level with object durations", async () => {
      mockClient.post.mockResolvedValue({ result: { id: "q-1", status: "sent" }, message: "Quote sent" });

      await TurboQuote.sendQuote("q-1", {
        ccEmails: ["admin@example.com"],
        remindersEnabled: true,
        reminderDelay: { value: 3, unit: "days" },
        reminderInterval: { value: 12, unit: "hours" },
        maxReminders: 5,
        expirationEnabled: true,
        expireAfter: { value: 30, unit: "days" },
        expirationWarning: { value: 3, unit: "days" },
        expirationWarningInterval: { value: 1, unit: "days" },
      });

      const body = mockClient.post.mock.calls[0][1];

      // Flat at the top level, never nested under "schedule".
      expect(body).not.toHaveProperty("schedule");

      // Native booleans / number — not stringified.
      expect(body.remindersEnabled).toBe(true);
      expect(body.expirationEnabled).toBe(true);
      expect(body.maxReminders).toBe(5);

      // Durations are OBJECTS, not JSON strings (this is a JSON endpoint).
      expect(body.reminderDelay).toEqual({ value: 3, unit: "days" });
      expect(typeof body.reminderDelay).toBe("object");
      expect(body.reminderInterval).toEqual({ value: 12, unit: "hours" });
      expect(body.expireAfter).toEqual({ value: 30, unit: "days" });
      expect(body.expirationWarning).toEqual({ value: 3, unit: "days" });
      expect(body.expirationWarningInterval).toEqual({ value: 1, unit: "days" });

      // Unrelated send options ride alongside.
      expect(body.ccEmails).toEqual(["admin@example.com"]);
    });

    it("omits every schedule key when the caller sets none, so the org defaults apply", async () => {
      mockClient.post.mockResolvedValue({ result: { id: "q-1", status: "sent" }, message: "Quote sent" });

      await TurboQuote.sendQuote("q-1", { ccEmails: ["admin@example.com"] });

      const body = mockClient.post.mock.calls[0][1];
      for (const key of SCHEDULE_KEYS) {
        expect(body).not.toHaveProperty(key);
      }
    });

    // `false` and `0` are meaningful, not "unset" — dropping them would silently fall back to the
    // org default, the opposite of what the caller asked for.
    it("preserves the meaningful zeros: maxReminders:0 and expirationEnabled:false", async () => {
      mockClient.post.mockResolvedValue({ result: { id: "q-1", status: "sent" }, message: "Quote sent" });

      await TurboQuote.sendQuote("q-1", { maxReminders: 0, expirationEnabled: false });

      const body = mockClient.post.mock.calls[0][1];
      expect(body.maxReminders).toBe(0);
      expect(body.expirationEnabled).toBe(false);
    });
  });

  describe("sendQuoteWithDeliverable — JSON send path", () => {
    it("carries the schedule FLAT alongside the deliverable fields", async () => {
      mockClient.post.mockResolvedValue({ result: { id: "q-1", status: "sent" }, message: "Sent", documentId: "doc-2" });

      await TurboQuote.sendQuoteWithDeliverable("q-1", {
        deliverableId: "del-1",
        mergePosition: "end",
        remindersEnabled: true,
        reminderDelay: { value: 2, unit: "days" },
        expirationEnabled: false,
      });

      const body = mockClient.post.mock.calls[0][1];
      expect(body).not.toHaveProperty("schedule");
      expect(body.deliverableId).toBe("del-1");
      expect(body.remindersEnabled).toBe(true);
      expect(body.reminderDelay).toEqual({ value: 2, unit: "days" });
      expect(body.expirationEnabled).toBe(false);
    });
  });

  describe("createAndSend — schedule rides on the send step", () => {
    it("emits the flat schedule on the /send request body", async () => {
      const quote = { id: "q-1", name: "Enterprise License", status: "draft" };
      mockClient.post
        .mockResolvedValueOnce({ result: quote, message: "Quote created successfully" })
        .mockResolvedValueOnce({ result: { ...quote, status: "sent" }, message: "Sent" });

      await TurboQuote.createAndSend({
        name: "Enterprise License",
        companyId: "c-1",
        contactId: "ct-1",
        send: {
          remindersEnabled: true,
          maxReminders: 0,
          reminderDelay: { value: 1, unit: "days" },
        },
      });

      const postCalls = mockClient.post.mock.calls;
      expect(postCalls[0][0]).toBe("/v1/quotes");
      expect(postCalls[1][0]).toBe("/v1/quotes/q-1/send");

      const sendBody = postCalls[1][1];
      expect(sendBody).not.toHaveProperty("schedule");
      expect(sendBody.remindersEnabled).toBe(true);
      expect(sendBody.maxReminders).toBe(0);
      expect(sendBody.reminderDelay).toEqual({ value: 1, unit: "days" });
    });
  });
});
